### PR TITLE
support heap glib tcache in dmh family

### PR DESCRIPTION
**Pull request by morbith-dqtz**
fix the whole patch, sorry about last PR

### CODE REVIEW

File: libr/core/linux_heap_glibc.c [hunk 1]
```diff
@@ -968,6 +1043,30 @@ static void GH(print_heap_segment)(RCore *core, GH(RHeap_MallocState) *main_aren
...
+		#if __GLIBC_MINOR__ > 25
+		GH(RHeapTcache) *tcache = R_NEW0 (GH(RHeapTcache));
+		(void)r_io_read_at (core->io, brk_start + MALLOC_ALIGNMENT, (ut8 *)tcache, sizeof ( GH(RHeapTcache) ));
+		for(int i=0; i<TCACHE_MAX_BINS; i++){
+			if(tcache->counts[i]>0x0){
+				if ( (ut64)tcache->entries[i]-MALLOC_ALIGNMENT == (ut64)prev_chunk ){
+					is_free = true;
+					break;
+				}
+				if(tcache->counts[i]>0x1){
```
Comment (radare) [inline libr/core/linux_heap_glibc.c]: missing spaces

File: libr/core/linux_heap_glibc.c [hunk 2]
```diff
@@ -968,6 +1043,30 @@ static void GH(print_heap_segment)(RCore *core, GH(RHeap_MallocState) *main_aren
 			}
 		}
 
+		#if __GLIBC_MINOR__ > 25
+		GH(RHeapTcache) *tcache = R_NEW0 (GH(RHeapTcache));
+		(void)r_io_read_at (core->io, brk_start + MALLOC_ALIGNMENT, (ut8 *)tcache, sizeof ( GH(RHeapTcache) ));
+		for(int i=0; i<TCACHE_MAX_BINS; i++){
+			if(tcache->counts[i]>0x0){
```
Comment (radare) [inline libr/core/linux_heap_glibc.c]: add spaces around > and just use 0, not 0x0

File: libr/core/linux_heap_glibc.c [hunk 3]
```diff
@@ -968,6 +1043,30 @@ static void GH(print_heap_segment)(RCore *core, GH(RHeap_MallocState) *main_aren
...
+				if ( (ut64)tcache->entries[i]-MALLOC_ALIGNMENT == (ut64)prev_chunk ){
+					is_free = true;
+					break;
+				}
+				if(tcache->counts[i]>0x1){
+					tcache_fd = (ut64)tcache->entries[i];
+					for(int n=1;n < tcache->counts[i];n++){
```
Comment (radare) [inline libr/core/linux_heap_glibc.c]: spacings after ; and around =

File: libr/core/linux_heap_glibc.c [hunk 4]
```diff
@@ -968,6 +1043,30 @@ static void GH(print_heap_segment)(RCore *core, GH(RHeap_MallocState) *main_aren
...
+						if ( (ut64)tcache_tmp-MALLOC_ALIGNMENT == (ut64)prev_chunk ) {
+							is_free = true;
+							break;
+						}
+						tcache_fd = (ut64)tcache_tmp;
+					}
+				}
+			}
+		}
+		#endif
```
Comment (radare) [inline libr/core/linux_heap_glibc.c]: all preprocessor lines should start at the begining of the line, don't indent them

File: libr/core/linux_heap_glibc.c [hunk 5]
```diff
@@ -983,8 +1082,16 @@ static void GH(print_heap_segment)(RCore *core, GH(RHeap_MallocState) *main_aren
 			if (!(cnk->size & 1)) {
 				PRINT_GA ("[free]");
 			} else {
+				#if __GLIBC_MINOR__ > 25
+				if(is_free) {
```
Comment (radare) [inline libr/core/linux_heap_glibc.c]: missing space before (

**Comment by XVilka** [comment]
What I don't like in this approach - it is compile-time configured. Imho, should be configured in runtime, not via the defines. For example, in case there will be a need to run `dmh` commands on a remote machine via gdb:// remote. I met the same problem with TCC (C types parsing code), where type sizes and defines were dependent on the compile time from the platform it was compiled on. While radare2 can target platforms different than it was compiled on. And I made it dependent on `e asm.bits` and `e asm.arch` instead, so you can change it on the fly. Thus I think the same approach should be applied to heap parsing as well.

**Pull request closed without merging**